File: services/admin-service/app/study_scope.py
"""The Usage and pipeline pages' study filter: one study, or several.

`study_id` on those endpoints takes one id or a comma-separated list --
the Usage page's "My studies" is every study the viewer belongs to, so
the figures are about work they can act on (UX: the page opened on every
study on the platform, the test studies included)."""
import uuid

from fastapi import HTTPException


def study_ids(value) -> list[str] | None:
    """None when there is no filter; else the ids, validated and
    normalised. A bad id is a 422 that names it."""
    if value is None:
        return None
    if isinstance(value, uuid.UUID):
        return [str(value)]
    parts = value if isinstance(value, (list, tuple, set)) else str(value).split(",")
    out: list[str] = []
    for part in parts:
        part = str(part).strip()
        if not part:
            continue
        try:
            out.append(str(uuid.UUID(part)))
        except ValueError:
            raise HTTPException(status_code=422, detail=f"'{part}' is not a valid study id") from None
    return out or None


def study_uuids(value) -> list[uuid.UUID] | None:
    ids = study_ids(value)
    return [uuid.UUID(i) for i in ids] if ids else None

File: admin-ui/src/pages/usage/studyScope.ts
/** The Usage page's study filter: all studies, "My studies" (every study
 * the viewer is a member of), or one study. It opens on My studies when
 * the viewer has any -- the page opened on every study on the platform,
 * the test studies included, and its "Act now" was about work the viewer
 * can't change -- and remembers the last choice in this browser. */

export const MY_STUDIES = "mine";
const REMEMBER_KEY = "vl.usage.study";

/** The studies I'm a member of or created, still in the list (a trashed
 * one isn't), in its order. */
export function myStudyIds(memberships: { study_id: string }[], studies: { id: string }[], created: string[] = []): string[] {
  const mine = new Set([...memberships.map((m) => m.study_id), ...created]);
  return studies.filter((s) => mine.has(s.id)).map((s) => s.id);
}

/** What the API's `study_id` gets: "" (no filter), one id, or "a,b,c". */
export function scopeOf(choice: string, mine: string[]): string {
  return choice === MY_STUDIES ? mine.join(",") : choice;
}

/** The remembered choice while it still makes sense, else My studies
 * when there are any, else all studies. */
export function initialChoice(remembered: string | null, mine: string[], studyIds: string[]): string {
  if (remembered === "") return "";
  if (remembered === MY_STUDIES && mine.length > 0) return MY_STUDIES;
  if (remembered && studyIds.includes(remembered)) return remembered;
  return mine.length > 0 ? MY_STUDIES : "";
}

export function rememberedChoice(): string | null {
  try {
    return localStorage.getItem(REMEMBER_KEY);
  } catch {
    return null;
  }
}

export function rememberChoice(choice: string): void {
  try {
    localStorage.setItem(REMEMBER_KEY, choice);
  } catch {
    // a blocked storage just means it isn't remembered
  }
}
